### PR TITLE
Force ppc64le builds to use gcc 13

### PR DESCRIPTION
**Pull request by djhoese**
Replaces #137 

Update: This PR was originally just for testing to find out why ppc64le builds were producing inconsistent results (ex. `cs2cs`). The problem was whittled down to gcc 12 being a problem. Rather than using the older gcc 11 (which works fine), this PR now changes the compiler to be gcc 13.

I've also added a test to the recipe that should fail if things go wrong again in a future version of the compiler.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by djhoese** [comment]
I made some guess as to how to get ppc64le to build on travis. Let me know (if the builds themselves don't) how this could/should theoretically be done. As mentioned in the other PR and issues, it seems there may be errors in the emulated ppc64le so we'll try native on travis.

**Comment by djhoese** [comment]
Hm the output still shows the invalid result:

```
363301.33	546152.98 0.00

363301.33	546152.98 0.00

589057.58	-3.54 0.00

569722.40	4268814.27 0.00

569722.40	4268814.27 0.00
```

(the -3.54 in the above is bad).

I don't know enough about travis but I *think* the configuration I did is telling it to use ppc64le natively...right?

Edit:

```
hostname: 707dbe23-d8d4-4c61-a985-31c94d3bb919@75129.lxd-ppc64le-travis-ci-production-1-worker2-london
```

**Comment by xylar** [comment]
> I don't know enough about travis but I think the configuration I did is telling it to use ppc64le natively...right?

That's my understanding, though my expertise is also somewhat limited.

**Comment by djhoese** [comment]
Note: Windows is failing just because I was lazy setting the debug env vars.

**Comment by djhoese** [comment]
Well...that's what I get for copying a patch from upstream `main`. Fixing up now...

**Comment by djhoese** [comment]
@rouault Sorry to pull you into this PR, but it might just be easier to discuss this this way: I've added your test from https://github.com/OSGeo/PROJ/pull/3923 as a patch to this feedstock. It fails on the travis ppc64le build. Conda-forge's CI scripts and build process is much more complicated than PROJ's travis setup. For reference:

PROJ:
https://app.travis-ci.com/github/OSGeo/PROJ/jobs/611673341

conda-forge:
https://app.travis-ci.com/github/conda-forge/proj.4-feedstock/jobs/611746723

The conda-forge build should be "native" but it does run through docker compared to PROJ's direct execution. @xylar let me know if there is some further option you can think of to run even more directly on the native ppc64le. I assume docker is smart enough to realize that it doesn't need to emulate the architecture at all (ppc running on ppc). The worker/build information shows that they are both running ppc64le.

Otherwise @rouault, conda-forge is still using:

https://github.com/conda-forge/proj.4-feedstock/blob/e622434c8fb2dd7134d582451d163fba7fab3943/recipe/define_OLD_BUGGY_REMQUO.patch

Do you imagine that this could be causing these ppc issues?

**Comment by rouault** [comment]
> Otherwise @rouault, conda-forge is still using:
> 
> https://github.com/conda-forge/proj.4-feedstock/blob/e622434c8fb2dd7134d582451d163fba7fab3943/recipe/define_OLD_BUGGY_REMQUO.patch
> 
> Do you imagine that this could be causing these ppc issues?

no, that's totally unrelated.

**Comment by rouault** [comment]
> Conda-forge's CI scripts and build process is much more complicated than PROJ's travis setup.

another difference is that PROJ's travis setup was using Ubuntu Bionic (18.04), so a much older build toolchain than conda-forge builds. I've updated https://github.com/OSGeo/PROJ/pull/3923 to use Jammy (22.04)

**Comment by rouault** [comment]
Tests also pass on Jammy: https://app.travis-ci.com/github/OSGeo/PROJ/builds/266655965 (the final failure is unrelated. The test suite success status is at https://app.travis-ci.com/github/OSGeo/PROJ/builds/266655965#L789)

**Comment by djhoese** [comment]
@xylar or other conda-forge expert: Is there an existing script/configuration for running linux builds natively (no docker)? I don't see any here:

https://github.com/conda-forge/conda-smithy/blob/main/conda_smithy/templates/run_docker_build.sh.tmpl

So I'm trying to hack my own just to get *something* working in this feedstock that passes the ppc64le test. Then I'll work from there to figure out what's causing the differences between here and upstream tests (docker build itself, conda-forge dependencies, etc).

**Comment by xylar** [comment]
> Is there an existing script/configuration for running linux builds natively (no docker)?

@djhoese, I don't think so, but I'm ready to be corrected by a conda-forge expert.

**Comment by djhoese** [comment]
@conda-forge-admin restart ci

**Comment by djhoese** [comment]
Ok that took way longer than expected, but here we have a failure without docker:

https://app.travis-ci.com/github/conda-forge/proj.4-feedstock/builds/266691392

@rouault main things that jump out to me when skimming the PROJ build log compared to your last travis job:

1. Newer compiler and dependency versions in the conda-forge build ("C compiler identification is GNU 12.3.0" versus 11.3.0). sqlite3 3.43 versus 3.37, libtiff 4.6 versus 4.3, curl 8.4 versus 7.81.
2. `- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed` in conda-forge, Success in proj. The conda-forge build does end up finding threads in "pthread" it says. I'm not familiar with what this check is doing when it checks libc, then pthreads, then pthread.

I could try limiting some of the versions of dependencies and see if that changes anything.

**Comment by rouault** [comment]
> I could try limiting some of the versions of dependencies and see if that changes anything.

I would tend to suspect the compiler change. If there was a systematic issue with the versions of libtiff/libsqlite3/libcurl, we'd observe it on x86_64 too (assuming here we don't have an arch-specific regression in those components, but they likely have no ppc64 assembly code). Something to check if changing compiler version is not easy would be to keep gcc 12.3, but do a -O0 build as compiler regressions tend to be on optimized builds

File: recipe/build.sh [hunk 1]
```diff
@@ -8,7 +8,8 @@ else
     EXE_SQLITE3=${BUILD_PREFIX}/bin/sqlite3
 fi
 
-cmake ${CMAKE_ARGS} \
+echo "Existing cmake args: ${CMAKE_ARGS}"
+cmake ${CMAKE_ARGS} -O0 \
       -D CMAKE_BUILD_TYPE=Release \
```
Comment (rouault) [inline recipe/build.sh]: I don't think cmake accepts -O0. That's a gcc flag. Try instead:
```suggestion
cmake ${CMAKE_ARGS} \
      -D CMAKE_BUILD_TYPE=Debug \
```

Comment (djhoese) [reply recipe/build.sh]: Yeah I woke up, read your comment, made a change thinking `CMAKE_FLAGS` was `CFLAGS`, and then realized the mistake later. I defined CFLAGS and CXXFLAGS and still got the failure. I'll remove those and change the build type now.

**Comment by djhoese** [comment]
After this build with the Debug release, I'll try changing the compiler and dependency versions.

File: recipe/build.sh (hunk 1 above)
Comment (djhoese) [reply recipe/build.sh]: Still failing. Now I'll try lowering compiler and dependency versions.

**Comment by djhoese** [comment]
Nobody panic! We have a passing ppc64le build! Ok let's try newer compilers as I'm scared to continue debugging if the compiler is the problem...

**Comment by djhoese** [comment]
Ah and I'll do it after this build, but I can probably turn on the original docker-based linux builds too and see how those go.

**Comment by djhoese** [comment]
Darn it. It's the compilers.

**Comment by djhoese** [comment]
Good news: Docker builds aren't the problem.
Good news: It isn't the dependencies.
Bad news: It's something about the compilers.

Ideas for next debugging steps (@xylar @rouault)? I had tried `-O0` with CFLAGS/CXXFLAGS and from what I could tell this was added correctly. I also had and still have the "Debug" build type for cmake.

**Comment by rouault** [comment]
> Ideas for next debugging steps

you're not going to like it... (and it would require to have sustained access to a powerful machine): build gcc from source and do a git bisect session on it between 11.3 and 12.3, building PROJ at each iteration...
Why do we care about ppc64le by the way...?

**Comment by djhoese** [comment]
> Why do we care about ppc64le by the way...?

:smile: I've been asking myself the same question, but assumed someone pushed the conda-forge core folks to build their ppc64le packages for them. I'm curious who the audience is.

And I'll see what I can do about the gcc stuff.

**Comment by djhoese** [comment]
I'm trying gcc "13" now which should be 13.2.0. "11" was good, "12" was bad. I'm not sure I can be more specific in the conda-forge config, but there are other versions available:

https://anaconda.org/conda-forge/gcc_linux-ppc64le/files

This might be the "cheap" option for narrowing down specific versions that are broken. If running 13 suddenly "just works" then maybe we can just get the recipe updated to use that and we ignore this...until it happens again.

Edit: GCC 12 mentions this caveat which is a little scary:

>An ABI incompatibility between C and C++ when passing or returning by value certain aggregates containing zero width bit-fields has been discovered on various targets. As mentioned in [PR102024](https://gcc.gnu.org/PR102024), since the [PR42217](https://gcc.gnu.org/PR42217) fix in GCC 4.5 the C++ front-end has been removing zero width bit-fields from the internal representation of the aggregates after the layout of those aggregates, but the C front-end kept them, so passing e.g. struct S { float a; int : 0; float b; } or struct T { float c; int : 0; } by value could differ between C and C++. Starting with GCC 12 the C++ front-end no longer removes those bit-fields from the internal representation and per clarified psABI some targets have been changed, so that they either ignore those bit-fields in the argument passing by value decisions in both C and C++, or they always take them into account. x86-64, ARM and AArch64 will always ignore them (so there is a C ABI incompatibility between GCC 11 and earlier with GCC 12 or later), PowerPC64 ELFv2 and S/390 always take them into account (so there is a C++ ABI incompatibility, GCC 4.4 and earlier compatible with GCC 12 or later, incompatible with GCC 4.5 through GCC 11). RISC-V has changed the handling of these already starting with GCC 10. As the ABI requires, MIPS takes them into account handling function return values so there is a C++ ABI incompatibility with GCC 4.5 through 11. For function arguments on MIPS, refer to [the MIPS specific entry](https://gcc.gnu.org/gcc-12/changes.html#mips_zero_width_fields). GCC 12 on the above targets will report such incompatibilities as warnings or other diagnostics unless -Wno-psabi is used.

**Comment by djhoese** [comment]
GCC 13 works! @xylar any rules for forcing gcc to a newer version than the default that conda-smithy gave? Or maybe that is just a separate migration? I changed it in `.ci_support` which I'm guessing is a no-no. Is there another way to do it? `conda-forge.yml`? `recipe/meta.yaml`?

**Comment by djhoese** [comment]
@ocefpaf same question for you if you're around and know the answer ^

**Comment by ocefpaf** [comment]
> @ocefpaf same question for you if you're around and know the answer ^

I'm quite removed from the compiler choices in conda-forge and maybe @isuruf can comment more. While I believe it is possible we need to be careful with compatibility.

With that said, I think I already said this but I'll say it again, we don't really make ppc64le a priority. I'm OK with just dropping it going forward and re-adding when a new compiler is available everywhere.

**Comment by djhoese** [comment]
I don't have any personal stake in ppc support so I don't think I should have a say in the outcome. I was just annoyed that something that seemed to be working before was now not working.

I do think this failure is an important one. @rouault what is your opinion on this test (or one like it) being included in PROJ's tests? It doesn't seem to be conda-forge specific anymore, but rather gcc 12 as the problem. I'd also be fine cleaning up my patch in this repository to only have the test here and/or update the shell-level `cs2cs` test to fail when it gets the wrong result.

Thanks @isuruf!

**Comment by rouault** [comment]
> what is your opinion on this test (or one like it) being included in PROJ's tests?

why not, but it needs some care as we try to make PROJ's test not to rely too much on the network (some builds might have curl disabled). If that's indeed a compiler bug, it is hard to correlate that with functional tests...

**Comment by djhoese** [comment]
Will the test you added in https://github.com/OSGeo/PROJ/pull/3923 silently skip if network isn't available or does it cause the test to fail in those environments? I'm thinking what I might do is do a simple shell-level test in conda-forge for `$(cs2cs ...) == $(cs2cs ...)`.

**Comment by rouault** [comment]
> or does it cause the test to fail in those environments?

it would fail

**Review by xylar — APPROVED** [review]
I can't say my proj expertise is deep but this fix seems very reasonable and I think we want to get it in as soon as possible.

**Comment by ocefpaf** [comment]
Awesome debugging work @djhoese! And thanks to all that helped and reviewed this one!!

**Comment by dopplershift** [comment]
Thanks for the hard work digging in @djhoese !

**Pull request merged**